lib/spack/spack/solver/asp.py

```diff
@@ -971,6 +971,66 @@ def _model_has_cycles(self, models):
         return cycle_result.unsatisfiable
 
 
+class ReusableSpecsByHash(collections.abc.Mapping):
+    """Mapping to contain reusable specs keyed by DAG hash.
+
+    The mapping is ensured to be consistent, i.e. if a spec in the mapping has a dependency with
+    hash X, it is ensured to be the same object in memory as the spec keyed by X.
+    """
+
+    def __init__(self) -> None:
+        self.data: Dict[str, spack.spec.Spec] = {}
+
+    def __getitem__(self, dag_hash: str) -> spack.spec.Spec:
+        return self.data[dag_hash]
+
+    def add(self, spec: spack.spec.Spec) -> bool:
+        """Adds a new concrete spec to the mapping. Returns True if the spec was just added,
+        False if the spec was already in the mapping.
+
+        Args:
+            spec: spec to be added
+
+        Raises:
+            ValueError: if the spec is not concrete
+        """
+        if not spec.concrete:
+            msg = (
+                f"trying to store the non-concrete spec '{spec}' in a container "
+                f"that only accepts concrete"
+            )
+            raise ValueError(msg)
+
+        dag_hash = spec.dag_hash()
+        if dag_hash in self.data:
+            return False
+
+        # Unify objects in the container
+        for edge in reversed(spack.traverse.traverse_edges_topo([spec], direction="children")):
```

Comment:
why edges?

Reply:
fyi: if the goal is to early exit on sub-dag added before, don't use topo order, cause it has to traverse the full dag to order things :)

Reply:
> if the goal is to early exit on sub-dag added before

No, the goal is to unify object for subdags added before - similarly to a roundtrip to-from DB. That unifies the 2 Python objects with the same DAG we have in memory.

Reply:
Can you do early exit though? It doesn't make sense to iterate over everything and skip each item from all subdags seen before.

Reply:
I mean, you must do this. Your current implementation effectively copies a database in O(n^2) time.

Reply:
Actually, `ReusableSpecsByHash` is rather wasteful, it always copies everything.

By design `daghash(x) == daghash(y) => id(x) == id(y)` for all `x` and `y` from the same database. So the issue is just about merging databases and adding specs from say spec.json files.

So, why not duplicate only specs that occur in multiple databases? There's already a dictionary of dag hashes per db.